Logit scale, changes in LogitLocator and LogitFormatter

jb-leger opened the pull request:
## PR Summary

I had a lot of bugs on the labelling on `logit` scale. For example, the following code:

```python
import numpy as np 
import scipy as sp 
import scipy.stats 
import matplotlib.pyplot as plt 
 
s = np.linspace(-5,5,10000) 
s2 = np.linspace(-10,10,10000) 
lims = [(1e-10,1-1e-10),(1e-5,1-1e-5),(11e-2,15e-2)] 
for k,lim in enumerate(lims): 
    fig,ax = plt.subplots() 
    ax.set_yscale('logit') 
    ax.plot(s,sp.stats.norm.cdf(s)) 
    for i in range(9): 
        ax.plot(s2,sp.stats.t(i+1).cdf(s2)) 
    ax.grid() 
    ax.set_ylim(*lim) 
    fig.savefig(f'/tmp/g{k}.png')
```                   

### Actual behavior

I obtain for the very large zoom:

![g0](https://user-images.githubusercontent.com/24868098/59163971-960d5980-8b07-11e9-9082-d5ce7bdf143b.png)

For normal zoom:

![g1](https://user-images.githubusercontent.com/24868098/59163973-a9202980-8b07-11e9-8472-9c5955b0ee6a.png)

For zoomed axis:

![g2](https://user-images.githubusercontent.com/24868098/59164017-3794ab00-8b08-11e9-9d14-eaff9726543d.png)

### Behavior on the proposed branch

All this graphs are obtained with the same code as above.

I change the locator to allow different behavior, on large zoom, some decade are not proposed on majors, but on minors. No minor ticks used for subdivision of decades. Label are present only on major ticks. I obtain:

![f0](https://user-images.githubusercontent.com/24868098/59164059-b5f14d00-8b08-11e9-8e4b-27ed4cf8d0cb.png)

On normal zoom, all decades are shown on major, minor are indicating subdecade. Most of the time only major ticks are labelled, but when only a few of major are present, a spaced subset of minors can be labelled. I obtain :

![f1](https://user-images.githubusercontent.com/24868098/59164118-15e7f380-8b09-11e9-966b-52fe6a70700c.png)

On a very zoomed graph, the logiscale is kind-of linear, so the `MaxNLocator` is used (in particular, my LogitLocator inherit from MaxNLocator to allows fallback to MaxNLocator in this case, I obtain:

![f2](https://user-images.githubusercontent.com/24868098/59164136-619a9d00-8b09-11e9-993b-92d2c6db68c3.png)

Also, two little things:
 - for ticks labelling, the precision of format is adaptative, and depends of the diff between the labelled tick and neighbors,
 - formater have `use_oveline` method, which allow use the survival notation (`\overline{x}` mean `1-x`). Not enabled by default.

Should solve #13698.

## PR Checklist

I putted my comments on each item.

- [X] Has Pytest style unit tests
   - I tried to test the most critical behaviors of my code.
- [X] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
   - Except for some new parameters, I did not add features.
- [ ] Documentation is sphinx and numpydoc compliant
   - I tried, but I did not test, and I don't know how to test that.
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
   - I don't know if this change is _major_. I think it is not.
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way
   - No backward incompatible API changes.

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jb-leger commented:
I force push to add some more tests (because codedev was complaining) and I rebased on `master` to resolve conflicts.

jb-leger commented:
Ok, finish! I add all the test I can add. Sorry for the noise of many force-push. It was only to keep a clean history.

jb-leger commented:
I push some change, if I have a good understanding of @anntzer comments in #14545.

 - I add the `minor` property in `LogitLocator` to keep the API compat.
 - all my new attributes was already private.
 - I switch at most as possible argument in keyword-only.

anntzer commented:
Thanks :)
From 30,000 feet the PR looks nice, but it's also quite a mouthful so it may take a while to properly review, just letting you know.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,188 @@ class LogitFormatter(Formatter):
+                precision = int(np.floor(-np.log10(diff) + exponent + 0.9))
+            if precision < 1:
+                precision = 1
+            s = (r"%%.%df\cdot10^{%%d}" % precision) % (value, exponent)
```
you can use asterisks as width specifiers (items 4 and 5 of https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting) to avoid having to nest format calls.

jb-leger replied:
done. Thanks, I did'nt know this specifiers.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,188 @@ class LogitFormatter(Formatter):
-                s = '$1-10^{{{:.0f}}}$'.format(np.log10(1-x))
+                s = r"1-%s" % s
+        elif x < 0.1:
+            exponent = int(np.floor(np.log10(x)))
```
you could use math.floor (here and below), which returns an int, to avoid the additional int() call.

jb-leger replied:
done. Thanks.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,188 @@ class LogitFormatter(Formatter):
-        else:  # x > 0.99
-            if is_decade(1-x):
-                s = '$1-10^{{{:.0f}}}$'.format(np.log10(1-x))
+                s = r"1-%s" % s
```
we're mostly using f-strings these days, although I'm fine with % for cases where there's a bunch of additional braces to escape.

jb-leger replied:
changed. Thanks.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,188 @@ class LogitFormatter(Formatter):
+        elif x > 0.9:
+            exponent = int(np.floor(np.log10(1 - x)))
+            value = (1 - x) * 10 ** (-exponent)
+            if len(self.locs) < 2:
```
some code looks duplicated between the x<0.1 and x>0.9 branches, it's a bit complex so may be worth factoring out?

jb-leger replied:
Factorized with a `.format_small_value`. Thanks for notice.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2079,13 +2242,13 @@ def decade_up(x, base=10):
 
 
-def is_decade(x, base=10):
+def is_decade(x, base=10, tol=None):
```
keyword-only for sure

jb-leger replied:
done. Sorry for that.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2138,8 +2301,10 @@ def _decade_greater(x, base):
 
-def is_close_to_int(x):
-    return abs(x - np.round(x)) < 1e-10
+def is_close_to_int(x, tol=None):
```
keyword only

jb-leger replied:
Done. Sorry for that.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2502,64 +2667,118 @@ def view_limits(self, vmin, vmax):
+        MaxNLocator.__init__(self, nbins=nbins, steps=[1, 2, 5, 10])
 
-    def set_params(self, minor=None):
+    def set_params(self, *, minor=None, **kwargs):
```
minor can't become kwonly, that breaks backcompat (though see cbook._make_keyword_only)

jb-leger replied:
Done. Sorry for that, I missed `minor` was previously present.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,188 @@ class LogitFormatter(Formatter):
+                dnext = np.concatenate(((0,), diff))
+                dprev = np.concatenate((diff, (0,)))
+                criterion = np.min((dnext, dprev), axis=0) + 0.1 * (
+                    dnext + dprev
```
the linebreak and continuation look a bit weird.

jb-leger replied:
It was formatted by `black`. I agree, a bit weird. Anyway, this is outdated by changing `np.min((…),axis=0)` to `np.minimum(…)`.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,188 @@ class LogitFormatter(Formatter):
+                diff = np.diff(-np.log(1 / alocs - 1))
+                dnext = np.concatenate(((0,), diff))
+                dprev = np.concatenate((diff, (0,)))
+                criterion = np.min((dnext, dprev), axis=0) + 0.1 * (
```
np.minimum(dnext, dprev) I think?

jb-leger replied:
exact. Modified. Thanks.

jb-leger commented:
Thanks @anntzer for you review. I take all you comments, I mark them as resolved (sorry for that if I didn't have to do this).

For readability I did'nt rebase for squashing commits, I can squash subsets of commits if needed.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -239,14 +240,120 @@ def test_set_params(self):
+                assert (len(major_ticks) - 1) * 5 < len(minor_ticks)
+            else:
+                # subsample
+                assert len(major_ticks) + len(minor_ticks) == len(
```
here too black's formatting is... less than great (FWIW I would use a backslash continuation, though it's just a suggestion)

jb-leger replied:
You are right, the second implies the first. Fixed.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +778,181 @@ def test_LogFormatter_call(self, val):
+            y = TestLogitFormatter.logit_deformatter(fx)
+            assert np.abs(x - y) < 1e-12
+
+    decade_test = (
```
decade_test = sorted(...)

jb-leger replied:
Fixed.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +778,181 @@ def test_LogFormatter_call(self, val):
+                TestLogitFormatter.logit_deformatter(fx)
+        else:
+            y = TestLogitFormatter.logit_deformatter(fx)
+            assert np.abs(x - y) < 1e-12
```
are the tolerance values here and below (1e-10) carefully determined, or should these basically just use e.g. pytest.approx/np.isclose/something else that sets the tolerance for you?

jb-leger replied:
I need more precise comparison than usual, because I comparing value near 0 or near 1. I fix that by using usal tolerance from `np.isclose` or `np.testing.assert_almost_equal` by doing the comparison in logit scale.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +778,181 @@ def test_LogFormatter_call(self, val):
+                sx1 = formatter(x + 1 / N)
+                d = TestLogitFormatter.logit_deformatter(
+                    sx1
+                ) - TestLogitFormatter.logit_deformatter(sx)
```
Again, this would be better formatted e.g. as
```
d = (TestLogit...(sx1)
     - TestLogit...(sx))
```
(or a bunch of other variants -- but you probably got my point that black isn't the universal solution :-))

jb-leger replied:
Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,181 @@ class LogitFormatter(Formatter):
+        """
+        self._use_overline = use_overline
+
+    def label_minor(self, minor):
```
Could this be basically folded into `set_minor_threshold(0)`? (not sure it's a nice API though)

jb-leger replied:
Not exactly,

 - if `_minor` is `True`, indicates that this formatter is used for minor ticks. Therefore, some ticks (spaced ticks as LogFormatter sublabels) are labelled.
 - if `_minor` is `False`, indicates that this formatter is used for major ticks. Therefore, all ticks are labelled.

`set_minor_threshold(0)` implies when `_minor` is `True` that no minor ticks is labelled.

I understand the difficulty. I thing the `label_minor` can be deleted. The `_minor` attr should be set only at the init.

I propose to delete the method `label_minor`. What do you think?

anntzer replied:
sounds reasonable.

jb-leger replied:
Ok, I set the default value in both.

jb-leger replied:
Done.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -239,14 +240,120 @@ def test_set_params(self):
+        """
+        expected_ticks = np.sort(
+            np.concatenate(
+                (expected_first_ticks, (1 / 2,), 1 - expected_first_ticks)
```
Something like
```
        expected = sorted([*expected_first, 1 / 2, *(1 - expected_first)])
```
would fit in a single line (it's not clear what "first" refers to, btw).

jb-leger replied:
Changed. (But it does'nt fit in one line (or maybe you expect I remove the `_ticks` in variable name, I prefer keeping a long name to have in mind that I manipulate ticks, it is not a strong preference).

"first" refers to ticks <.5, I build all the ticks using first ticks. I you have a better idea, let me know. I am not very happy with this name.

anntzer replied:
low_ticks perhaps? at least low/high maps more clearly to <0.5/>0.5 than first/second, I think.
The exact formatting doesn't matter too much but it was just me once again not liking what black did :)

jb-leger replied:
Changed first to low.

Me too I do not like what black do, but what I like with black, it is that I can stop thinking of hundreds way to format the same thing.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +778,181 @@ def test_LogFormatter_call(self, val):
+            comp = match["comp"] is not None
+            mantissa = (
+                float(match["mant"])
+                if (match["mant"] is not None and len(match["mant"]) > 0)
```
that's just `if match["mant"]`?  same below.

jb-leger replied:
Yes. It was incremental work, I start with `match[…]` is not None` and I add constraints…
Changed.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +778,181 @@ def test_LogFormatter_call(self, val):
+    @pytest.mark.parametrize("x", 1 / (1 + np.exp(-np.linspace(-7, 7, 10))))
+    def test_variablelength(self, x):
+        """
+        The format lenght should change depending on the neighbor labels.
```
typo: length

jb-leger replied:
Fixed. Sorry for that.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +778,181 @@ def test_LogFormatter_call(self, val):
+        form = mticker.LogitFormatter()
+        for x in locs:
+            fx = form.format_data_short(x)
+            if len(fx) > 2 and fx[0:2] == "1-":
```
fx.startswith("1-")

jb-leger replied:
Nice, I completely forgot this method! Thanks.
Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,181 @@ class LogitFormatter(Formatter):
+
+    def use_overline(self, use_overline):
+        r"""
+        Switch display mode with overline for labelling p>1/2
```
add end dots to docstrings (throughout)

jb-leger replied:
Done.

anntzer commented:
May also be nice to add or modify an example in https://matplotlib.org/gallery/index.html#ticks-and-spines to showcase the new functionalities this adds.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,181 @@ class LogitFormatter(Formatter):
+                self._labelled.update(locs)
             else:
-                s = '$1-{:.5f}$'.format(1-x)
+                alocs = np.array(locs)
```
This block is a bit obscure and could use a comment explaining what's happening.

jb-leger replied:
Done. I do not mark this thread as solve. It depends if you think it is clearer now.

anntzer replied:
It's not clear what criterion (or even diff) below refers to.  In particular, is 0.1 really 1/base, or is it just some small value intended for tiebreaking? (in which case you could use tuple ordering instead?)

jb-leger replied:
It was only a tiebreak. I change it to tuple ordering, and I change the variable name to highlight the logic. Same as previous, I keep this thread open.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2079,13 +2235,13 @@ def decade_up(x, base=10):
 
 
-def is_decade(x, base=10):
+def is_decade(x, base=10, *, tol=None):
```
perhaps name this rtol? (well it's not *strictly* a relative tolerance, but it's closer to a relative tolerance than to an absolute one)

jb-leger replied:
It is a absolute tolerance in logspace, then, for me, it is a absolute tolerance.

Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2138,8 +2294,10 @@ def _decade_greater(x, base):
 
-def is_close_to_int(x):
-    return abs(x - np.round(x)) < 1e-10
+def is_close_to_int(x, *, tol=None):
```
atol (as opposed to rtol)
Not sure we need to support tol=None, could just do tol=1e-10 as default.

jb-leger replied:
I think we need `None` as default, for the `is_decade` function, or we have to set the default value in both function.

anntzer replied:
I think setting the default in both places is just fine.  (Note that if you were to write docs for these helpers (not requesting it for this PR) you'd have to document the default in both places anyways...)

jb-leger replied:
Fixed, I set the default value in both. And I change tol to atol.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2502,64 +2660,119 @@ def view_limits(self, vmin, vmax):
+            if numideal > nbins:
+                # to many ideal ticks, subsampling ideals for major ticks, and
+                # take others for minor ticks
+                subsampling_factor = int(np.ceil(numideal / nbins))
```
math.ceil

jb-leger replied:
fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,177 @@ class LogitFormatter(Formatter):
+            precision = 0
+        else:
+            diff = np.sort(np.abs(locs - x))[1]
+            precision = math.floor(-np.log10(diff) + exponent + 0.9)
```
same comment here for the 0.9 as for the 0.1 earlier: is that really 1-1/base, or should that whole thing just use math.ceil?

jb-leger replied:
`math.ceil` is what I wanted to do, exempt log10 have imprecision effects. I change using ceil when it is a real float and round when it is near a int.

Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,177 @@ class LogitFormatter(Formatter):
+            diff = np.sort(np.abs(locs - x))[1]
+            precision = math.floor(-np.log10(diff) + exponent + 0.9)
+        if precision < 1:
+            precision = 0
```
`precision = max(precision, 0)`? possibly folded into the else: clause above?  (dunno, just asking)

jb-leger replied:
Completely equivalent for me. In my opinion, it is more understandable without a max.

anntzer replied:
sure

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2502,64 +2654,119 @@ def view_limits(self, vmin, vmax):
+        self._minor = minor
+        MaxNLocator.__init__(self, nbins=nbins, steps=[1, 2, 5, 10])
+
+    @cbook._make_keyword_only("3.2", "minor")
```
is there a reason for the API change here, and making `minor` a property?  (yes, I know, I suggested doing similar stuff in the docs PR, but that's for new APIs; we could also migrate older APIs to newer design but here it just lengthens the PR and makes it harder to review (unless there's a good reason for the change that I missed))

jb-leger replied:
As I rewrite competely the LogitLocator, I decide use `_minor` as private. I add the property after to have the compat with the previous locator. I do not know if it is a good reason.

anntzer replied:
Do you actually want to deprecate `.minor`?  If so, that could be done in a separate PR.
(IOW the changes here are pretty orthogonal to the rest, so given the big size of the current PR, would benefit from being split out.)

jb-leger replied:
It is why I add a property without deprecate it, to easily add the deprecation on the getter and the setter without modifying the rest of the new code.

jb-leger replied:
I remove the deprecation of this args. Deprecation will be the topic of a future PR.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2502,64 +2654,119 @@ def view_limits(self, vmin, vmax):
+        # linscale: ... 1e-3 1e-2 1e-1 1/2 1-1e-1 1-1e-2 1-1e-3 ...
+        # b-scale : ... -3   -2   -1   0   1      2      3      ...
+        def ideal_ticks(x):
+            return (
```
I think the whole expression (without parentheses) fits in a single line? (I don't mind breaking it e.g. after each `else` for clarity, but here the linebreaks achieve little.)

jb-leger replied:
You were right. Originally I wrote this expression as a lambda fun, but my linter was complaining. Fixed.

jb-leger commented:
> May also be nice to add or modify an example in https://matplotlib.org/gallery/index.html#ticks-and-spines to showcase the new functionalities this adds.

@anntzer, I remove the NullFormatter() from the pyplot tutorial, and in the `examples/scales/scales.py`, since it is not necessary now (and it mask the behavior of the `LogitFormatter` with `minor=True`).

Also I add a new example `demo_logit.py`, in scales category.

anntzer commented on examples/scales/logit_demo.py:
```diff
@@ -0,0 +1,57 @@
+        axs[i, j].legend()
+        axs[i, j].grid()
+
+# First line, logiscale, with standard notation
```
typo: logit (again below)

jb-leger replied:
Thanks.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,177 @@ class LogitFormatter(Formatter):
+                # we do not have a lot of minor ticks, so only few decades are
+                # displayed, then we choose some (spaced) minor ticks to label.
+                alocs = np.array(locs)
+                diff = np.diff(-np.log(1 / alocs - 1))
```
Not really familiar with logit formulae but what's this log(1/x-1) that appears again and again?

jb-leger replied:
In math, `logit(x) = log(x/(1-x))`, we can write in base `e` or in any base. For all this computation knowing the base is useless.

For computing, to avoid numerical error, some of us (as me) prefer use the equivalent formula: `logit(x) = -log(1/x-1)` we have smaller errors.

jb-leger replied:
For example:

```
In [1]: import numpy as np                                                                                                              

In [2]: logit1 = lambda x: -np.log(1/x -1)                                                                                              

In [3]: logit2 = lambda x: np.log(x/(1-x))                                                                                              

In [4]: expit = lambda x: 1/(1+np.exp(-x))                                                                                              

In [5]: logit1(expit(10)) - 10                                                                                                          
Out[5]: -1.2967404927621828e-13

In [6]: logit2(expit(10)) - 10                                                                                                          
Out[6]: 9.698908343125368e-13
```

anntzer replied:
OK, now I think I got it: you want to put your minor ticks as far as possible from the major ticks (taking the spacing between the major ticks as secondary tiebreaker), is that correct?  If so please add a comment explaining that, I've been staring at these formulae for way too long...
Also does that even work once you start adding more than one minor tick per space? wouldn't you pick two "central" ticks in that case, rather than having them at ~1/3 each?
Finally (again assuming I understood what you're doing...) note that the rest of the code already assumes base 10, in which case... wouldn't the minor ticks always appear in the same order anyways?  Do you need to redo this ordering computation every time, or can you just hard-code the minor ticks order?  (Still not sure I fully understood what's happening here.)

jb-leger replied:
You have the idea, except the formatter do not know the space from the next major ticks but only the space from the next minor ticks, which is not a big issue, because the nearest minor ticks is the nearest for most the case (except for .9, which is anywar near .8).

Yes, precomputed ticks could do the job for most case, but in fact we have a lot of case to distinguish:
 - many decades < .1
 - some decades <.1 and other is .5
 - some decades <.1 and other is .9
 - .1 and .5
 - …

So I prefer use a automatic computation than precompute in many case. LogFormatter have precomputed values, but it doesn't have many case to distinguish.

Please note this computation is done only one time for all values, not for each.

anntzer replied:
Thanks.
Given how much time it took me to understand the computation here, there should probably be some comments in the code explaining what's going on...

jb-leger replied:
I put more comments, let me know.

anntzer replied:
thanks, just a couple of typos but looks much better :)

jb-leger replied:
Thanks, it was not a very easy point to explain.

anntzer commented on examples/scales/logit_demo.py:
```diff
@@ -0,0 +1,57 @@
+
+import numpy as np
+import scipy as sp
+import scipy.stats
```
right now the docs build independently of scipy; it's not really worth adding scipy as a dependency just for this (you can implement whatever cdf formula you want inline?)

jb-leger replied:
I cant. For inline this I need the hypergeometric function which is in scipy.special.

But it is only a example, I can change the distribution used in example.

anntzer replied:
please do

jb-leger replied:
Done. I choose simpler distributions. I was prefering using students distrib, but it is only because it is my self usage of the logit scale.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,182 @@ class LogitFormatter(Formatter):
+        self._minor_number = minor_number
+
+    def set_locs(self, locs):
+        self.locs = locs
```
should you cast locs to an array here?

jb-leger replied:
All other formatter which set locs do not cast it. But other do nothing with locs (as `LogFormatter`). I can do that I want, it is not specified by the base class `Formatter`.

I will cast it.

jb-leger replied:
Done.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,182 @@ class LogitFormatter(Formatter):
-                s = '$1-{:.5f}$'.format(1-x)
+                # we do not have a lot of minor ticks, so only few decades are
+                # displayed, then we choose some (spaced) minor ticks to label.
+                alocs = np.array(locs)
```
if you cast locs to an array at the top you don't need to do it again here

jb-leger replied:
Done.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,182 @@ class LogitFormatter(Formatter):
+            s = self._format_small_value(x, self.locs)
+        elif x > 0.9:
+            s = self._one_minus(
+                    self._format_small_value(1-x, [1-y for y in self.locs])
```
the second argument is just `1-self.locs`?

jb-leger replied:
If cast as you propose above, yes.

jb-leger replied:
Done.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,182 @@ class LogitFormatter(Formatter):
+                precision = 1
+            else:
+                diff = np.sort(np.abs(self.locs - x))[1]
+                precision = math.floor(-np.log10(diff) + 0.9)
```
again, math.ceil?

jb-leger replied:
Same as previous, round if close to int else math.ceil. I will change that.

jb-leger replied:
done.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,182 @@ class LogitFormatter(Formatter):
+            if len(self.locs) < 2:
+                precision = 1
+            else:
+                diff = np.sort(np.abs(self.locs - x))[1]
```
would be nice to have a comment explaining where this "second smallest difference" comes from

jb-leger replied:
it is only because the first smallest difference is zero, it is the difference between the ticks and itself. I will add a comment.

jb-leger replied:
done.

anntzer commented on lib/matplotlib/scale.py:
```diff
@@ -667,14 +667,19 @@ class LogitScale(ScaleBase):
+
+        use_overline: bool (default: False)
+            indicate the usage of survival notation (\overline{x}) in place of
+            standard notation (1-x) for probability close to one.
```
How standard is this notation? Do you have an example of its use (e.g. on a publication)?  Just curious...

jb-leger replied:
This notation is very common in extreme value theory for survival function (1-cdf()). Generally we note $F$ the cdf, and $\overline{F}$ the survival. [For example, in the introduction](https://hal.archives-ouvertes.fr/hal-00942134/document). The use of the notation for numerical value exists, but it is not so common (personally I dislike this notation for numerical values).

 - For sets: \(A \cup \overline{A} = U\), with \(U\) the universe, the most common.
 - For survival function: \(F(x) + \overline{F}(x) = 1\), common, I use it.
 - For probability: \( p + \overline{p} = 1\), rare, I do not use it.

anntzer replied:
If it's rare and you don't like it, do we really want to even implement it? (remember, any feature we implement means more code to maintain, more bug reports, etc.) I don't feel very strongly about it either way, though.

jb-leger replied:
Some colleagues of mine prefer use this notation in internal documents, but they change notation for publication. I ask them, having this notation is not a major issue. The concerned code is not large, it is deduplicated by the internal `_one_minus`, therefore I suggest keep this functionality. But I agree to remove it if asked.

anntzer replied:
that's fine, you can leave it.

anntzer commented on examples/scales/logit_demo.py:
```diff
@@ -0,0 +1,60 @@
+"""
```
The rendered example is at https://22432-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/scales/logit_demo.html#sphx-glr-gallery-scales-logit-demo-py; you may want to play with e.g. figsize a bit...
Also, looking at the example I wonder whether $\frac{1}{2}$ is that nice, perhaps 1/2 would look better?

jb-leger replied:
yes. There is a way to build the html locally?

For one half, I prefer $\frac{1}{2}$, but it is a personal pov, I can put the choice in a option.

anntzer replied:
Yes, see https://matplotlib.org/devel/documenting_mpl.html#building-the-docs.
I'll let the next reviewer (we need two to merge) decide on whether to insist on an API for 1/2 vs. \frac...

jb-leger replied:
I think both choice should be proposed, but we still have to decide the default choice.

anntzer replied:
It's your PR, you get to pick, unless the other reviewer agrees with me that 1/2 is better :)

jb-leger replied:
I have added a option to choose the string for one half. It is illustrated in the example. I can change the default choice.

anntzer commented:
Before I forget: I think this is a nice enough improvement that it also warrants a whatsnew entry (in doc/users/next_whats_new).

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,192 @@ class LogitFormatter(Formatter):
-                s = '$1-{:.5f}$'.format(1-x)
+                # we do not have a lot of minor ticks, so only few decades are
+                # displayed, then we choose some (spaced) minor ticks to label.
+                # Only minor ticks are known, we assume it is suffisant to
```
sufficient

jb-leger replied:
oops, my spellcheck was disabled. 5 mistakes in 3 sentences.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,192 @@ class LogitFormatter(Formatter):
+                # choice which ticks are displayed.
+                # For each ticks we compute the distance beteewn the ticks and
+                # the previous, and between the ticks and the next one. Ticks
+                # with smallest minimum are choosen. As teebreak, the ticks
```
tiebreak

jb-leger replied:
sorry for the noise.

anntzer commented on lib/matplotlib/scale.py:
```diff
@@ -667,14 +667,31 @@ class LogitScale(ScaleBase):
+
+        one_half : str (default: r"\frac{1}{2}")
+            the string used for ticks formatter to represent 1/2. If None,
+            "0.5" is used.
```
Again I wouldn't bother with supporting None here; but if you do want to support None, then it should match the default (that's the typical Matplotlib behavior), so "r\frac{1}{2}".

jb-leger replied:
None is not supported now. Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,208 @@ class LogitFormatter(Formatter):
+            if len(self.locs) < 2:
+                precision = 1
+            else:
+                # we look for the nearest ticks from loc (except the tick at
```
looks like some of this code is duplicated in _format_small_value?  can it be merged with it?  (dunno)

jb-leger replied:
I changed `_format_small_value` to `_format_value` and deduplicate this section. Thanks for notice. Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,208 @@ class LogitFormatter(Formatter):
+        elif x < 0.1:
+            s = self._format_small_value(x, self.locs)
+        elif x > 0.9:
+            s = self._one_minus(
```
now fits in one line

jb-leger replied:
Fixed.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -2502,64 +2685,116 @@ def view_limits(self, vmin, vmax):
+        numideal = bsup - binf - 1
+        if numideal >= 2:
+            # have 2 or more wanted ideal ticks, so use them as major ticks
+            if numideal > nbins:
```
I think(?) you can perhaps get rid of this close and take advantage of the fact that if numideal <= nbins, then subsampling_factor = math.ceil(numideal / nbins) == 1 so there's no subsampling.

Also it *may* be easier to follow if you split apart the major and minor cases first?  Right now it's not that easy to follow the major/minor cases.

Just some suggestions, nothing compulsory.

jb-leger replied:
Maybe it because I did this code, I understand what you suggest, but the resulting code seem more difficult to understand. But I wrote this code, I am partial.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -238,15 +239,132 @@ def test_set_params(self):
 
+class LogitHelper:
+    @staticmethod
+    def isclose_in_logitspace(x, y):
```
perhaps make these toplevel (private) helpers (`_is_close_in_logitspace`) or at least don't repeat "logit" in the name? (`_LogitSpace.is_close`, `_LogitSpace.assert_almost_equal`)?  The repetition of "logit" makes lines unduly long below.

jb-leger replied:
I rename the class as private. In remove the repetition for logit in the function names.

For testing I am generally less astringent for private/public object. It why I didn't put this helper as private.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -671,6 +789,185 @@ def test_LogFormatter_call(self, val):
+                    TestLogitFormatter.logit_deformatter(sx1)
+                    - TestLogitFormatter.logit_deformatter(sx)
+                )
+                assert d > 0
```
`0 < d < 2/N` is handled just fine by pytest

jb-leger replied:
Fixed. Thanks for notice.

anntzer commented:
I think the only missing thing is a whatsnew entry.

jb-leger commented:
I just add it. I dislike writing this type of text. I think improvements are needed.

anntzer commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+Improvements in Logit scale ticker and formatter
+-----------------------------------------------
```
```
/home/circleci/project/doc/users/next_whats_new/logit_scale_stuff.rst:2: WARNING: Title underline too short.

Improvements in Logit scale ticker and formatter
-----------------------------------------------
```

jb-leger replied:
I blame my spell checker , which correct the title without correction the underline.

Will be fixed.

jb-leger replied:
fixed, committed, ans squashed.

anntzer commented:
The whatsnew entry is mostly useful so that you're not the sole user of the feature :)

jb-leger commented:
Sorry for the delay, I am quite busy these days.

I squash commits in 10 commits, grouping by coherent subsets of changes.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -2079,13 +2254,13 @@ def decade_up(x, base=10):
 
 
-def is_decade(x, base=10):
+def is_decade(x, base=10, *, rtol=1e-10):
```
Should this be `atol` instead of `rtol` ?

jb-leger replied:
It is a absolute tolerance after log transform, but a relative tolerance before log transform, therefore, for me both are correct, so I follow @anntzer choice.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        """
+        Return a short formatted string representation of a number.
+        """
+        if value < 0.1:
```
Can you add a comment about why these threshold were picked?  It seems odd that we get the `1-N` format only from `0.9 < v < 1`

jb-leger replied:
It is only because I dislike format as `1.31e-1' or '1-2.31e-1'. More precisely thresholds choosen to have a exponent less or equal than `-2`.

Comment added.

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -2502,64 +2677,116 @@ def view_limits(self, vmin, vmax):
-        """Return the locations of the ticks."""
-        vmin, vmax = self.axis.get_view_interval()
-        return self.tick_values(vmin, vmax)
+    @property
```
Does this need to be a property?

jb-leger replied:
Yes, to keep to compatibility with the old `LogitLocator` which has a `minor` attribute.

tacaswell commented:
Thanks for your work on this @jb-leger !

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+Improvements in Logit scale ticker and formatter
+------------------------------------------------
+
+Introduced in version 1.5, the logit scale didn't have appropriate ticker and
```
```suggestion
Introduced in version 1.5, the logit scale didn't have an appropriate ticker and
```

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+------------------------------------------------
+
+Introduced in version 1.5, the logit scale didn't have appropriate ticker and
+formatter. Previously, location of ticks was not zoom dependent, too many label
```
```suggestion
formatter. Previously, the location of ticks was not zoom dependent, too many labels
```

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+
+Introduced in version 1.5, the logit scale didn't have appropriate ticker and
+formatter. Previously, location of ticks was not zoom dependent, too many label
+was displayed implying overlapping which break readability, and label formatting
```
```suggestion
were displayed causing overlapping which broke readability, and label formatting
```

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+Introduced in version 1.5, the logit scale didn't have appropriate ticker and
+formatter. Previously, location of ticks was not zoom dependent, too many label
+was displayed implying overlapping which break readability, and label formatting
+was not precision adaptive.
```
```suggestion
did not adapt to precision.
```

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+was displayed implying overlapping which break readability, and label formatting
+was not precision adaptive.
+
+Starting from this version, the locator have near the same behavior as the
```
```suggestion
Starting from this version, the logit locator has nearly the same behavior as the
```

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+was not precision adaptive.
+
+Starting from this version, the locator have near the same behavior as the
+locator for the log scale or the same behavior as the locator for the linear
```
```suggestion
locator for the log scale or the linear
```

QuLogic commented on doc/users/next_whats_new/logit_scale_stuff.rst:
```diff
@@ -0,0 +1,13 @@
+locator for the log scale or the same behavior as the locator for the linear
+scale, depending on used zoom. The number of ticks is controlled. Some minor
+labels are displayed adaptively as sublabels in log scale. Formatting is adapted
+for probabilities and the precision is adaptive depending on the scale.
```
```suggestion
for probabilities and the precision is adapts to the scale.
```

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -667,14 +667,30 @@ class LogitScale(ScaleBase):
           values beyond ]0, 1[ can be masked as invalid, or clipped to a number
           very close to 0 or 1
+
+        use_overline: bool (default: False)
```
```suggestion
        use_overline : bool, default: False
```

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -667,14 +667,30 @@ class LogitScale(ScaleBase):
+            indicate the usage of survival notation (\overline{x}) in place of
+            standard notation (1-x) for probability close to one.
+
+        one_half : str (default: r"\frac{1}{2}")
```
```suggestion
        one_half : str, default: r"\frac{1}{2}"
```

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -667,14 +667,30 @@ class LogitScale(ScaleBase):
           very close to 0 or 1
+
+        use_overline: bool (default: False)
+            indicate the usage of survival notation (\overline{x}) in place of
```
```suggestion
            Indicate the usage of survival notation (\overline{x}) in place of
```

QuLogic commented on lib/matplotlib/scale.py:
```diff
@@ -667,14 +667,30 @@ class LogitScale(ScaleBase):
+            standard notation (1-x) for probability close to one.
+
+        one_half : str (default: r"\frac{1}{2}")
+            the string used for ticks formatter to represent 1/2.
```
```suggestion
            The string used for ticks formatter to represent 1/2.
```

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -3,6 +3,7 @@
 import numpy as np
 from numpy.testing import assert_almost_equal, assert_array_equal
 import pytest
+import re
```
This import belongs with `warnings`.

jb-leger replied:
I do not understand what you mean by *belongs with `warnings`*. Doesn't mean I do not have to import `re` ?

QuLogic replied:
According to [PEP8](https://www.python.org/dev/peps/pep-0008/#imports), imports are grouped by their origin: standard library, then third party, then local imports, separated by a blank line. `re` and `warnings` are standard library and should be grouped together.

jb-leger replied:
Thanks for explanation.

Changed. (I close this thread).

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        r"""
+        Parameters
+        ----------
+        use_overline : bool (default: False)
```
```suggestion
        use_overline : bool, default: False
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        Parameters
+        ----------
+        use_overline : bool (default: False)
+            if x > 1/2, with x = 1-v, indicate if x should be displayed as
```
```suggestion
            If x > 1/2, with x = 1-v, indicate if x should be displayed as
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            if x > 1/2, with x = 1-v, indicate if x should be displayed as
+            $\overline{v}$. The default is to display $1-v$.
+
+        one_half : str (default: r"\frac{1}{2}")
```
```suggestion
        one_half : str, default: r"\frac{1}{2}"
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            $\overline{v}$. The default is to display $1-v$.
+
+        one_half : str (default: r"\frac{1}{2}")
+            the string used to represent 1/2.
```
```suggestion
            The string used to represent 1/2.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        one_half : str (default: r"\frac{1}{2}")
+            the string used to represent 1/2.
+
+        minor : bool (default: False)
```
```suggestion
        minor : bool, default: False
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            the string used to represent 1/2.
+
+        minor : bool (default: False)
+            indicate if the formatter is formatting minor ticks or not.
```
```suggestion
            Indicate if the formatter is formatting minor ticks or not.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        minor : bool (default: False)
+            indicate if the formatter is formatting minor ticks or not.
+            Basically minor ticks are not labelled, except when only few ticks
+            are provided, the most espaced ticks are labelled. See others
```
There's a typo here, but not sure what "most espaced" should be replaced with (as "most spaced" is also not clear to me.)

Also, others -> other

jb-leger replied:
I changed the `others`.

I reword the spacing between ticks.

I keep this thread open, I am not sure about my words.

jb-leger replied:
You are right for the idea, but it is not what the code is doing. The formatter for minor ticks only know the location for minor ticks and doesn't know the location of major ticks, and only minor ticks is used.

Due to behavior on LogitLocator, the only way to having few (<3) major ticks is equivalent to have a few (<25) minor ticks. So the code apply the threshold on the number of minor ticks.

Same idea for spacing, from the space between minor ticks and the LogitLocator behavior, making the decision from space between minor ticks is sufficient to have a nice behavior.

Therefore I think keeping the name ticks without qualifier is a solution. The other solution, is to use the minor qualifier all time and add another comment explaining why only minor ticks are used.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            are provided, the most espaced ticks are labelled. See others
+            parameters to change the default behavior.
+
+        minor_threshold : int (default: 25)
```
```suggestion
        minor_threshold : int, default: 25
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            parameters to change the default behavior.
+
+        minor_threshold : int (default: 25)
+            maximum number of locs for labelling some minor ticks. This
```
```suggestion
            Maximum number of locs for labelling some minor ticks. This
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            maximum number of locs for labelling some minor ticks. This
+            parameter have no effect if minor is False.
+
+        minor_number : int (default: 6)
```
```suggestion
        minor_number : int, default: 6
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+            parameter have no effect if minor is False.
+
+        minor_number : int (default: 6)
+            number of ticks which are labelled when the number of ticks is
```
```suggestion
            Number of ticks which are labelled when the number of ticks is
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+
+        Parameters
+        ----------
+        use_overline : bool (default: False)
```
```suggestion
        use_overline : bool, default: False
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        Parameters
+        ----------
+        use_overline : bool (default: False)
+            if x > 1/2, with x = 1-v, indicate if x should be displayed as
```
```suggestion
            If x > 1/2, with x = 1-v, indicate if x should be displayed as
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        r"""
+        Set the way one half is displayed.
+
+        one_half : str (default: r"\frac{1}{2}")
```
```suggestion
        one_half : str, default: r"\frac{1}{2}"
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        Set the way one half is displayed.
+
+        one_half : str (default: r"\frac{1}{2}")
+            the string used to represent 1/2.
```
```suggestion
            The string used to represent 1/2.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        Parameters
+        ----------
+        minor_threshold : int
+            maximum number of locs for labelling some minor ticks. This
```
```suggestion
            Maximum number of locations for labelling some minor ticks. This
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+
+    def set_minor_threshold(self, minor_threshold):
+        """
+        Set the threshold for labelling minors ticks
```
```suggestion
        Set the threshold for labelling minors ticks.
```

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1170,25 +1170,200 @@ class LogitFormatter(Formatter):
+        Parameters
+        ----------
+        minor_number : int
+            number of ticks which are labelled when the number of ticks is
```
```suggestion
            Number of ticks which are labelled when the number of ticks is
```

jb-leger commented:
Thanks @tacaswell for your review. Sorry for the delay, I was away any kind of computers.

jb-leger commented:
Thanks @QuLogic for all your comments and suggestions. It was the first time I use the functionnality of github for applying suggestions as commits.

jb-leger commented:
Rebased on master, to fix the errors implied by changes on CI (for example package rename, fixed by [this commit](https://github.com/matplotlib/matplotlib/commit/d1722d31f187f0c3991fdeb5d38f8bc7b0fb724b)).

I hope tests will pass now.

jb-leger commented:
Rebased to master for resolve conflict introduce by merge #14654.
